### DISCUSSION TITLE
Exporting and importing self designed components

### DISCUSSION THREAD

**Discussion by Joostth98** [mjhoptics/ray-optics#128] [Q&A] [answered]
Dear Micheal Hayford,

Is there a possibility to easily save an optical model as a sequential model and import it into another optical model? 
For example, I have modeled a telescope and I want to optimise the Oculair but I do this in a different notebook. And thereafter I would import it in my telescope notebook.

Kind regards,

Joost ter Haar

**Comment by mjhoptics** [comment, marked as answer]
Hello @Joostth98,
Thank you for posting this question. I've put together an example below to show how to save and combine lens models in ray-optics. Please let me know if you have further questions.
Thank you!
Mike Hayford

# Issue #128 - Combining files example

This example uses the add_from_file method of the OpticalModel to insert the lens in the file into the optical model. Any file type that can be imported by ray-optics (e.g. .zmx, .seq, .roa) can be used.

The example is in 3 parts:

 - Create an apochromatic triplet model and save it
 - Create an orthoscopic eyepiece model and save it
 - Using add_from_file, create a telscope model from the objective and eyepiece models.

The lens prescriptions are from the book [Telescopes, Eyepieces and Astrographs](https://shopatsky.com/products/telescopes-eyepieces-and-astrographs).


```python
%matplotlib inline
```


```python
isdark = False
```


```python
from rayoptics.environment import *
```

# Create an Apochromatic triplet


```python
opm = OpticalModel()
opm['system_spec'].title = 'Apo-Triplet, 150mm, f/8, Fig 6.22'
sm  = opm['seq_model']
osp = opm['optical_spec']
pm = opm['parax_model']
em = opm['ele_model']
pt = opm['part_tree']
```


```python
osp['pupil'] = PupilSpec(osp, key=['object', 'pupil'], value=150.0)
osp['fov'] = FieldSpec(osp, key=['object', 'angle'], value=0.35, flds=[0., 1.], is_relative=True)
osp['wvls'] = WvlSpec([("F", 0.5), ('d', 1.0), ('C', 0.5)], ref_wl=2)
```


```python
opm.radius_mode = True
sm.do_apertures = False

sm.gaps[0].thi=1e10

sm.add_surface([591.0786, 12, 'N-BK7', 'Schott', 80])
sm.add_surface([292.7895, 2, '', '', 80])
sm.add_surface([286.5319, 25, 'S-FPL53', 'Ohara', 80])
sm.add_surface([-469.3911, 2, '', '', 80])
sm.add_surface([-461.4804, 12, 'N-SK5', 'Schott', 80])
sm.add_surface([-1106.461, 1123.698, '', '', 80])
```


```python
opm.update_model()
```


```python
sm.gaps[-1].thi += pm.thi_ht_solve(pm.ax, -1, 0.)
opm.update_model()
```


```python
sm.list_model()
```

                  r            t        medium     mode   zdr      sd
      Obj:     0.000000  1.00000e+10       air             1      1.0000
        1:   591.078600      12.0000     N-BK7             1      80.000
        2:   292.789500      2.00000       air             1      80.000
        3:   286.531900      25.0000   S-FPL53             1      80.000
        4:  -469.391100      2.00000       air             1      80.000
        5:  -461.480400      12.0000     N-SK5             1      80.000
        6: -1106.461000      1173.91       air             1      80.000
      Img:     0.000000      0.00000                       1      1.0000



```python
pm.list_model()
```

               ax_ht        pr_ht       ax_slp       pr_slp         power          tau        index    type
     0:            0 -6.10873e+07      7.5e-09   0.00610873             0        1e+10      1.00000    dummy
     1:           75            0   -0.0652606   0.00610873  0.0008701421      7.92434      1.51432    transmit
     2:      74.4829    0.0484076     0.065578   0.00619376  -0.001756628            2      1.00000    transmit
     3:       74.614    0.0607951   -0.0483052   0.00610097   0.001526299      17.3933      1.43733    transmit
     4:      73.7738     0.166911    -0.117041   0.00594546  0.0009317036            2      1.00000    transmit
     5:      73.5397     0.178802   -0.0236272   0.00617258  -0.001270244      7.56528      1.58619    transmit
     6:       73.361       0.2255   -0.0624931   0.00605311  0.0005297907      1173.91      1.00000    transmit
     7:            0      7.33128   -0.0624931   0.00605311             0            0      1.00000    dummy



```python
layout_plt = plt.figure(FigureClass=InteractiveLayout, opt_model=opm, is_dark=isdark).plot()
```

![output_12_0](https://github.com/mjhoptics/ray-optics/assets/33294031/1ff90e97-b3d5-417b-92c8-38e329b235e2)

```python
opm.save_model('Apo-Triplet.roa')
```

# Define an Orthoscopic eyepiece, 10mm


```python
opm = OpticalModel()
sm  = opm['seq_model']
osp = opm['optical_spec']
pm = opm['parax_model']
em = opm['ele_model']
pt = opm['part_tree']
ar = opm['analysis_results']
```


```python
osp['pupil'] = PupilSpec(osp, key=['object', 'pupil'], value=1.0)
osp['fov'] = FieldSpec(osp, key=['object', 'angle'], value=22.5, flds=[0., 0.707, 1.], is_relative=True)
osp['wvls'] = WvlSpec([('F', 0.5), ('d', 1.0), ('C', 0.5)], ref_wl=1)
```


```python
opm.radius_mode = True

sm.gaps[0].thi=1e10

sm.add_surface([0, 7.244451])
sm.set_stop()
sm.add_surface([0, 3., 'N-SK16', 'Schott'])
sm.add_surface([-11.92452, .25])
sm.add_surface([11.29189, 4, 'N-SK16,Schott'])
sm.add_surface([-11.92452, 1, 'N-SF1,Schott'])
sm.add_surface([11.92452, 3, 'N-SK16,Schott'])
sm.add_surface([-60.94443, 4.864743])
```


```python
opm.update_model()
```


```python
sm.list_model()
```

                  r            t        medium     mode   zdr      sd
      Obj:     0.000000  1.00000e+10       air             1  4.1421e+09
     Stop:     0.000000      7.24445       air             1     0.50000
        2:     0.000000      3.00000    N-SK16             1      3.5007
        3:   -11.924520     0.250000       air             1      4.0570
        4:    11.291890      4.00000    N-SK16             1      4.3741
        5:   -11.924520      1.00000     N-SF1             1      4.2351
        6:    11.924520      3.00000    N-SK16             1      4.1492
        7:   -60.944430      4.86474       air             1      4.1194
      Img:     0.000000      0.00000                       1      3.7937



```python
listobj(osp)
```

    aperture: object epd; value=1.0
    field: object angle; value=22.5
    x,y=0.0 vlx= 0.000 vux= 0.000 vly= 0.000 vuy= 0.000
    y=0.707 vly= 0.000 vuy= 0.000 vlx= 0.000 vux= 0.000
    y=1.0 vly= 0.000 vuy= 0.000 vlx= 0.000 vux= 0.000
    central wavelength=  587.5618 nm
    wavelength (weight) =  486.1327 (0.500),  587.5618 (1.000)*,  656.2725 (0.500)
    focus shift=0.0
    



```python
pm.first_order_data()
```

    efl               10.01
    f                 10.01
    f'                10.01
    ffl             -0.7289
    pp1               9.277
    bfl               4.908
    ppk              -5.098
    pp sep            4.119
    f/#               10.01
    m            -1.001e-09
    red          -9.994e+08
    obj_dist          1e+10
    obj_ang            22.5
    enp_dist             -0
    enp_radius          0.5
    na obj            5e-11
    n obj                 1
    img_dist          4.908
    img_ht            4.145
    exp_dist         -132.4
    exp_radius        6.864
    na img         -0.04991
    n img                 1
    optical invariant       0.2071



```python
pm.list_model()
```

               ax_ht        pr_ht       ax_slp       pr_slp         power          tau        index    type
     0:            0 -4.14214e+09        5e-11     0.414214             0        1e+10      1.00000    dummy
     1:          0.5            0        5e-11     0.414214             0      7.24445      1.00000    transmit
     2:          0.5      3.00075        5e-11     0.414214             0      1.85138      1.62041    transmit
     3:          0.5      3.76762    -0.026014     0.218192    0.05202809         0.25      1.00000    transmit
     4:     0.493496      3.82217   -0.0531282   0.00819047    0.05494297      2.46851      1.62041    transmit
     5:     0.362349      3.84238   -0.0501822    0.0394303   -0.00813032     0.582289      1.71736    transmit
     6:     0.333128      3.86534   -0.0474738    0.0708568   -0.00813032      1.85138      1.62041    transmit
     7:     0.245236      3.99653   -0.0499702    0.0301724    0.01017993      4.86474      1.00000    transmit
     8:   0.00214389      4.14331   -0.0499702    0.0301724             0            0      1.00000    dummy



```python
layout_plt = plt.figure(FigureClass=InteractiveLayout, opt_model=opm, is_dark=isdark).plot()
```

    
![output_23_0](https://github.com/mjhoptics/ray-optics/assets/33294031/502a5100-ff45-4a1e-a2bb-e428e4165547)


```python
opm.save_model('orthoscopic eyepiece')
```

# Combine Objective and Eyepiece


```python
opm = OpticalModel()
sm  = opm['seq_model']
osp = opm['optical_spec']
pm = opm['parax_model']
em = opm['ele_model']
pt = opm['part_tree']
ar = opm['analysis_results']
```


```python
osp['pupil'] = PupilSpec(osp, key=['object', 'pupil'], value=100.0)
osp['fov'] = FieldSpec(osp, key=['object', 'angle'], value=0.35, flds=[0., 0.707, 1.], is_relative=True)
osp['wvls'] = WvlSpec([('F', 0.5), ('d', 1.0), ('C', 0.5)], ref_wl=1)
```


```python
opm.radius_mode = True

sm.gaps[0].thi=1e10
```

## Restore the objective using add_from_file()

First restore the objective and set the paraxial focal distance


```python
opm.add_from_file('Apo-Triplet.roa', t=1173.91)
```


```python
sm.gaps[-1].thi += pm.thi_ht_solve(pm.ax, -1, 0.)
opm.update_model()
```

## Restore the eyepiece


```python
opm.add_from_file('orthoscopic eyepiece.roa', t=7.24445)
```


```python
opm.update_model()
```


```python
sm.list_model()
```

                  r            t        medium     mode   zdr      sd
      Obj:     0.000000  1.00000e+10       air             1  6.1087e+07
        1:   591.078600      12.0000     N-BK7             1      50.013
        2:   292.789500      2.00000       air             1      49.661
        3:   286.531900      25.0000   S-FPL53             1      49.770
        4:  -469.391100      2.00000       air             1      49.441
        5:  -461.480400      12.0000     N-SK5             1      49.297
        6: -1106.461000      1173.81       air             1      49.217
        7:     0.000000      7.24445       air             1      7.3371
        8:     0.000000      3.00000    N-SK16             1      7.6830
        9:   -11.924520     0.250000       air             1      7.6886
       10:    11.291890      4.00000    N-SK16             1      4.3741
       11:   -11.924520      1.00000     N-SF1             1      4.2351
       12:    11.924520      3.00000    N-SK16             1      4.1492
       13:   -60.944430      7.24445       air             1      4.1194
      Img:     0.000000      0.00000                       1      7.3371


## Flip the eyepiece so that the pupil follows the eyepiece

Set the final distance to be at the paraxial exit pupil


```python
opm.flip(8,13)

sm.gaps[7].thi = 4.864743

sm.gaps[-1].thi += pm.thi_ht_solve(pm.pr, -1, 0.)
opm.update_model()
```


```python
sm.list_model()
```

                  r            t        medium     mode   zdr      sd
      Obj:     0.000000  1.00000e+10       air             1  6.1087e+07
        1:   591.078600      12.0000     N-BK7             1      50.013
        2:   292.789500      2.00000       air             1      49.661
        3:   286.531900      25.0000   S-FPL53             1      49.770
        4:  -469.391100      2.00000       air             1      49.441
        5:  -461.480400      12.0000     N-SK5             1      49.297
        6: -1106.461000      1173.81       air             1      49.217
        7:     0.000000      4.86474       air             1      7.3371
        8:    60.944430      3.00000    N-SK16             1      7.5920
        9:   -11.924520      1.00000     N-SF1             1      7.5959
       10:    11.924520      4.00000    N-SK16             1      7.7868
       11:   -11.291890     0.250000       air             1      7.6285
       12:    11.924520      3.00000    N-SK16             1      4.7682
       13:    -0.000000      8.05583       air             1      4.0726
      Img:     0.000000      0.00000                       1      10.864



```python
listobj(osp)
```

    aperture: object epd; value=100.0
    field: object angle; value=0.35
    x,y=0.0 vlx= 0.000 vux= 0.000 vly= 0.000 vuy= 0.000
    y=0.707 vly= 0.000 vuy= 0.000 vlx= 0.000 vux= 0.000
    y=1.0 vly= 0.000 vuy= 0.000 vlx= 0.000 vux= 0.000
    central wavelength=  587.5618 nm
    wavelength (weight) =  486.1327 (0.500),  587.5618 (1.000)*,  656.2725 (0.500)
    focus shift=0.0
    



```python
pm.first_order_data()
```

    efl           2.808e+05
    f             2.808e+05
    f'            2.808e+05
    ffl           3.368e+07
    pp1           3.396e+07
    bfl               -2333
    ppk          -2.832e+05
    pp sep       -3.424e+07
    f/#                2799
    m            -2.799e-05
    red          -3.573e+04
    obj_dist          1e+10
    obj_ang            0.35
    enp_dist             -0
    enp_radius           50
    na obj            5e-09
    n obj                 1
    img_dist          -2326
    img_ht             1710
    exp_dist          -2326
    exp_radius       0.4169
    na img       -0.0001787
    n img                 1
    optical invariant       0.3054



```python
pm.list_model()
```

               ax_ht        pr_ht       ax_slp       pr_slp         power          tau        index    type
     0:            0 -6.10873e+07        5e-09   0.00610873             0        1e+10      1.00000    dummy
     1:           50            0   -0.0437167   0.00610873  0.0008743339      7.91139      1.51680    transmit
     2:      49.6541    0.0483285    0.0439274   0.00619403  -0.001765091            2      1.00000    transmit
     3:       49.742    0.0607166   -0.0322398   0.00610106   0.001531245      17.3762      1.43875    transmit
     4:      49.1818      0.16673   -0.0782111   0.00594521  0.0009347225            2      1.00000    transmit
     5:      49.0254      0.17862   -0.0156249   0.00617324  -0.001276609       7.5513      1.58913    transmit
     6:      48.9074     0.225236   -0.0416654   0.00605332  0.0005324454      1173.81      1.00000    transmit
     7:            0       7.3307   -0.0416654   0.00605332             0      4.86474      1.00000    transmit
     8:    -0.202691      7.36015    -0.039602   -0.0688724    0.01017993      1.85138      1.62041    transmit
     9:     -0.27601      7.23264   -0.0418461   -0.0100688   -0.00813032     0.582289      1.71736    transmit
    10:    -0.300376      7.22677   -0.0442882    0.0486872   -0.00813032      2.46851      1.62041    transmit
    11:    -0.409702      7.34696   -0.0217779    -0.354977    0.05494297         0.25      1.00000    transmit
    12:    -0.415147      7.25821 -0.000178652    -0.732608    0.05202809      1.85138      1.62041    transmit
    13:    -0.415478      5.90188 -0.000178652    -0.732608             0      8.05583      1.00000    transmit
    14:    -0.416917  0.000118792 -0.000178652    -0.732608             0            0      1.00000    dummy



```python
layout_plt = plt.figure(FigureClass=InteractiveLayout, opt_model=opm, is_dark=isdark).plot()

```


![output_42_0](https://github.com/mjhoptics/ray-optics/assets/33294031/6eddbb3e-1f6c-4917-a07a-1bf6e651995e)



```python

```

**Reply by Joostth98** [reply]
Thanks for the quick response, very helpful!
